### PR TITLE
feat alter:  parallelize processing of parts in alter modify

### PR DESCRIPTION
**Pull request by ivan-v-kush**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category:
- Performance Improvement

Short description:
 * paralellize processing of parts in alter modify

### CODE REVIEW

File: dbms/src/Storages/StorageMergeTree.cpp [hunk 1]
```diff
@@ -221,14 +259,7 @@ void StorageMergeTree::alter(
...
 
-    auto parts = data.getDataParts({MergeTreeDataPartState::PreCommitted, MergeTreeDataPartState::Committed, MergeTreeDataPartState::Outdated});
-    auto columns_for_parts = new_columns.getAllPhysical();
-    std::vector<MergeTreeData::AlterDataPartTransactionPtr> transactions;
-    for (const MergeTreeData::DataPartPtr & part : parts)
-    {
-        if (auto transaction = data.alterDataPart(part, columns_for_parts, new_indices.indices, false))
-            transactions.push_back(std::move(transaction));
-    }
+    auto transactions = prepareAlterTransactions(new_columns, new_indices, 2 * getNumberOfPhysicalCPUCores());
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageMergeTree.cpp]: No idea, why we have 2 here.
Maybe introduce a setting (`max_alter_threads`, similar to `max_threads`)?
(This setting will make sense for default (system) profile.)

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

File: dbms/src/Storages/StorageMergeTree.cpp (hunk 1 above)
Comment (IvanKush) [reply dbms/src/Storages/StorageMergeTree.cpp]: hyperthreading? 2 threads by core

I copypasted from 
dbms/programs/copier/ClusterCopier.cpp
```
ThreadPool thread_pool(num_threads ? num_threads : 2 * getNumberOfPhysicalCPUCores());
```

Comment (IvanKush) [reply dbms/src/Storages/StorageMergeTree.cpp]: Should we introduce 1 threadpool for all ALTER operations? Not only for Modify column?

File: dbms/src/Storages/StorageMergeTree.h [hunk 2]
```diff
@@ -132,6 +132,9 @@ class StorageMergeTree : public ext::shared_ptr_helper<StorageMergeTree>, public
 
     BackgroundProcessingPool::TaskHandle background_task_handle;
 
+    std::vector<MergeTreeData::AlterDataPartTransactionPtr> prepareAlterTransactions(
+        const ColumnsDescription & new_columns, const IndicesDescription & new_indices, const size_t thread_pool_size);
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageMergeTree.h]: Still waiting for moving the number of threads to `max_alter_threads` setting (see `Settings.h`).

File: dbms/src/Storages/StorageMergeTree.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/src/Storages/StorageMergeTree.cpp]: Answered verbally.
(no need to have a single ThreadPool)

File: dbms/src/Storages/StorageMergeTree.cpp [hunk 3]
```diff
@@ -191,6 +193,42 @@ void StorageMergeTree::rename(const String & new_path_to_db, const String & /*ne
 }
 
 
+std::vector<MergeTreeData::AlterDataPartTransactionPtr> StorageMergeTree::prepareAlterTransactions(
+    const ColumnsDescription & new_columns, const IndicesDescription & new_indices, const size_t thread_pool_size)
+{
+    ThreadPool thread_pool(thread_pool_size);
```
Comment (alexey-milovidov) [inline dbms/src/Storages/StorageMergeTree.cpp]: Let's also cap the thread pool size to the number of data parts.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

**Comment by IvanKush** [comment]
@alexey-milovidov 

The problem is that we lock data part's mutex in threadpool's threads and unlock in the main thread

What should be done? Add supression in ThreadSanitizer? Or refactor the creating/commiting of  `AlterDataPartTransactionPtr`s?

**Description:**
Tsan reports:
```
WARNING: ThreadSanitizer: unlock of an unlocked mutex (or by a wrong thread) (pid=242)
    #0 pthread_mutex_unlock <null> (clickhouse+0x66cc82e)
    #1 std::__1::unique_lock<std::__1::mutex>::unlock() /usr/include/c++/v1/__mutex_base:270:11 (clickhouse+0xcd70063)
    #2 DB::MergeTreeData::AlterDataPartTransaction::clear() /build/obj-x86_64-linux-gnu/../dbms/src/Storages/MergeTree/MergeTreeData.h:237 (clickhouse+0xcd70063)

```

The problem is that we call `alterDataPart` in threadpool. In this func we create  AlterDataPartTransactionPtr, which constructor locks data part's `alter_mutex`. So mutex locks in some threadpool's threads.
```
AlterDataPartTransaction(DataPartPtr data_part_) : data_part(data_part_), alter_lock(data_part->alter_mutex) {}
```

But later in the main thread in func `alter` we call `transaction.commit()` which calls `clear` which unlocks  data part's mutex

```
 void clear()
{
    alter_lock.unlock();
    data_part = nullptr;
}
```

ThreadSanitizer doesn't like it

File: dbms/src/Storages/StorageMergeTree.cpp (hunk 3 above)
Comment (IvanKush) [reply dbms/src/Storages/StorageMergeTree.cpp]: done

File: dbms/src/Storages/StorageMergeTree.h (hunk 2 above)
Comment (IvanKush) [reply dbms/src/Storages/StorageMergeTree.h]: done

File: dbms/src/Storages/StorageMergeTree.cpp (hunk 1 above)
Comment (IvanKush) [reply dbms/src/Storages/StorageMergeTree.cpp]: done

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
It is only implemented for MergeTree, but not for ReplicatedMergeTree.

**Pull request merged**